cli_tools/google-osconfig-agent/config/config.go

```diff
@@ -37,19 +51,119 @@ var (
+	agentConfigMx.RLock()
+	defer agentConfigMx.RUnlock()
+	return *agentConfig
+}
+
+func enabled(s string) bool {
+	enabled, err := strconv.ParseBool(s)
+	if err != nil {
+		// Bad entry returns as not enabled.
+		return false
+	}
+	return enabled
+}
+
+type metadataJSON struct {
+	Instance instanceJSON
+	Project  projectJSON
+}
+
+type instanceJSON struct {
+	Attributes attributesJSON
+}
+
+type projectJSON struct {
+	Attributes attributesJSON
+}
+
+type attributesJSON struct {
+	OSInventoryEnabled string `json:"os-inventory-enabled"`
+	OSPatchEnabled     string `json:"os-patch-enabled"`
+	OSPackageEnabled   string `json:"os-package-enabled"`
+}
+
+func (c *config) setFromMetadata(md metadataJSON) {
+	// Check project first then instance as instance metadata overrides project.
+	if md.Project.Attributes.OSInventoryEnabled != "" {
+		c.osInventoryEnabled = enabled(md.Project.Attributes.OSInventoryEnabled)
+	}
+	if md.Project.Attributes.OSPatchEnabled != "" {
+		c.osPatchEnabled = enabled(md.Project.Attributes.OSPatchEnabled)
+	}
+	if md.Project.Attributes.OSPackageEnabled != "" {
+		c.osPackageEnabled = enabled(md.Project.Attributes.OSPackageEnabled)
+	}
+
+	if md.Instance.Attributes.OSInventoryEnabled != "" {
+		c.osInventoryEnabled = enabled(md.Instance.Attributes.OSInventoryEnabled)
+	}
+	if md.Instance.Attributes.OSPatchEnabled != "" {
+		c.osPatchEnabled = enabled(md.Instance.Attributes.OSPatchEnabled)
+	}
+	if md.Instance.Attributes.OSPackageEnabled != "" {
+		c.osPackageEnabled = enabled(md.Instance.Attributes.OSPackageEnabled)
+	}
+}
+
+// SetConfig sets the agent config.
+func SetConfig() error {
+	agentConfigMx.Lock()
+	defer agentConfigMx.Unlock()
+	agentConfig = &config{
+		osInventoryEnabled: osInventoryEnabled,
+		osPackageEnabled:   osPackageEnabled,
+		osPatchEnabled:     osPatchEnabled,
+		googetRepoFilePath: googetRepoFilePath,
+		zypperRepoFilePath: zypperRepoFilePath,
+		yumRepoFilePath:    yumRepoFilePath,
+		aptRepoFilePath:    aptRepoFilePath,
+	}
+
+	var md string
+	var err error
+	webError := 0
+	for {
+		md, err = metadata.Get("?recursive=true&alt=json")
+		if err == nil {
+			break
+		}
+		// Only log the third web error to avoid transient errors and
+		// not to spam the log on network failures.
+		if webError == 1 {
+			if urlErr, ok := err.(*url.Error); ok {
+				if _, ok := urlErr.Err.(*net.DNSError); ok {
+					return fmt.Errorf("DNS error when requesting metadata, check DNS settings and ensure metadata.internal.google is setup in your hosts file")
+				}
+				if _, ok := urlErr.Err.(*net.OpError); ok {
+					return fmt.Errorf("network error when requesting metadata, make sure your instance has an active network and can reach the metadata server")
+				}
+			}
+			return fmt.Errorf(err.Error())
+		}
+		webError++
+		time.Sleep(5 * time.Second)
+	}
+
+	var metadata metadataJSON
+	if err := json.Unmarshal([]byte(md), &metadata); err != nil {
+		return err
+	}
+
+	agentConfig.setFromMetadata(metadata)
```

Comment:
if you change `setFromMetadata()` to `buildFromMetadata()`, you can put the logic around locking the `agentConfig ` pointer here where you set it.